File: java/core/src/main/java/com/github/fernthedev/core/packets/Packet.java
package com.github.fernthedev.core.packets;

import com.github.fernthedev.core.encryption.codecs.AcceptablePacketTypes;
import lombok.Getter;
import lombok.NonNull;

import java.lang.reflect.Modifier;

public abstract class Packet implements AcceptablePacketTypes {
    private static final long serialVersionUID = -5039841570298012421L;

    @Getter
    protected final transient @NonNull String packetName;

    public static String getPacketName(Class<?> packet) {
        if (packet.equals(Packet.class)) throw new IllegalArgumentException("The class cannot be " + Packet.class.getName());

        if (!Packet.class.isAssignableFrom(packet)) throw new IllegalStateException("Packet " + packet.getName() + " must extend " + Packet.class.getName());

        if (Modifier.isAbstract(packet.getModifiers()) || packet.isInterface()) throw new IllegalArgumentException("The class cannot be abstract or interface.");

        if (!packet.isAnnotationPresent(PacketInfo.class)) throw new IllegalStateException("Packet " + packet.getName() + " must have a packet info annotation");

        return packet.getAnnotation(PacketInfo.class).name();
    }


    protected Packet() {
        if (!getClass().isAnnotationPresent(PacketInfo.class)) throw new IllegalStateException("Packet must have a packet info annotation");

        packetName = getClass().getAnnotation(PacketInfo.class).name();
    }
}

File: csharp/core/packets/Packet.cs
using com.github.fernthedev.lightchat.core.encryption;
using com.github.fernthedev.lightchat.core.util;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.Diagnostics;
using System.Diagnostics.CodeAnalysis;
using System.Text;
using System.Text.Json.Serialization;

namespace com.github.fernthedev.lightchat.core.packets
{
    public abstract class Packet : IAcceptablePacketTypes
    {
        protected Packet()
        {
            if (!Attribute.IsDefined(GetType(), typeof(PacketInfoAttribute))) throw new ArgumentException("Packet " + GetType().FullName + " must have a packet info annotation");

            _PacketName = ((PacketInfoAttribute) Attribute.GetCustomAttribute(GetType(), typeof(PacketInfoAttribute)))?.Name;
        }

        [field: NonSerialized]
        [Newtonsoft.Json.JsonIgnore]
        [JsonIgnore]
        public string _PacketName { get; }

        public static string getPacketName(Type packet)
        {

            if (packet == typeof(Packet)) throw new ArgumentException("The class cannot be " + typeof(Packet).FullName);

            if (!typeof(Packet).IsAssignableFrom(packet)) throw new ArgumentException("Packet " + packet.FullName + " must extend " + typeof(Packet).FullName);

            if (packet.IsAbstract || packet.IsInterface) throw new ArgumentException("The class cannot be abstract or interface.");

            if (!Attribute.IsDefined(packet, typeof(PacketInfoAttribute))) throw new ArgumentException("Packet " + packet.FullName + " must have a packet info annotation");

            return ((PacketInfoAttribute) Attribute.GetCustomAttribute(packet, typeof(PacketInfoAttribute)))?.Name;
        }
    }

    [AttributeUsage(AttributeTargets.Class)]
    public sealed class PacketInfoAttribute : Attribute
    {
        [NotNull]
        public string Name { get; }

        public PacketInfoAttribute(string name)
        {
            Debug.Assert(name != null, nameof(name) + " != null");
            Name = name;
        }
    }
}
